[HIP-0026] Plugin OCI installer

scottrigby opened the pull request:
Adds OCI installer to Helm 4 plugins. Also updates pkg/getter and pkg/registry for better re-use. These packages now allow downloading plugins as well as charts. No public SDK changes for these packages were needed in this implementation.

Part of implementing hip-0026. For full context, see:
- https://github.com/helm/helm/issues/31144

Depends on:
- #31146
Please review that PR first. After that PR is merged, will change this PR base branch to `main` and move out of draft.

Checklist:
- [x] This PR contains user facing changes
- [x] this PR contains unit tests
- [x] this PR has been tested for backwards compatibility

### Manual testing

Unit tests are comprehensive. You can also manually test with:

```console
% go run ./cmd/helm registry login ghcr.io
Login Succeeded

% go run ./cmd/helm plugin install oci://ghcr.io/scottrigby/example-cli:v0.2.0
Installed plugin: example-cli

% go run ./cmd/helm plugin list                                               
NAME			VERSION		TYPE		APIVERSION		SOURCE                                              
example-cli		0.1.0  		cli/v1		v1				https://github.com/scottrigby/h4-simple-plugin-types
```

Source code for example plugin: https://github.com/scottrigby/h4-example-plugin-types.

scottrigby committed:
Plugin OCI installer

Signed-off-by: Scott Rigby <scott@r6by.com>

TerryHowe commented on pkg/registry/plugin.go:
```diff
@@ -0,0 +1,176 @@
+	Ref            string
+	PluginName     string
+}
+
```
Given some greenfield project I would of tried to provide some generic content generator to pass to a generic oci client rather than the content having a client, but given the current code base, I think this make more sense.

TerryHowe commented on pkg/registry/oci.go:
```diff
@@ -0,0 +1,162 @@
+	"oras.land/oras-go/v2/registry/remote/auth"
+	"oras.land/oras-go/v2/registry/remote/credentials"
+)
+
```
why isn't this file named `generic_client.go`?

jessesimpson36 commented on pkg/cmd/plugin_install.go:
```diff
@@ -68,10 +88,39 @@ func (o *pluginInstallOptions) complete(args []string) error {
+		}
+
+		if o.username != "" || o.password != "" {
+			options = append(options, getter.WithBasicAuth(o.username, o.password))
```
Not that I have a hard opinion on this, but I'm wondering if these `getter.WithBasicAuth()` or other `With` function calls could contain the logic for `if o.username != "" || if o.password != ""`

Then this `newInstallerForSource` function would just be a registration of options, like

```
options = append(options, getter.WithTLSClientConfig(o.certFile, o.keyFile, o.caFile))
options = append(options, getter.WithInsecureSkipVerifyTLS(o.insecureSkipTLSverify))
options = append(options, getter.WithPlainHTTP(o.plainHTTP))
options = append(options, getter.WithBasicAuth(o.username, o.password))

```

scottrigby commented on internal/plugin/installer/oci_installer.go:
```diff
@@ -0,0 +1,229 @@
+
+// NewOCIInstaller creates a new OCIInstaller with optional getter options
+func NewOCIInstaller(source string, options ...getter.Option) (*OCIInstaller, error) {
+	ref := strings.TrimPrefix(source, fmt.Sprintf("%s://", registry.OCIScheme))
```
`newReference` is nice. looks like you made it private about 6 months ago https://github.com/helm/helm/pull/12690/commits/d2b94f62004e79864ec530989109cf0effd4aaae. Was it for a reason that's still relevant now?

If there's no objection to making reference struct and newReference func public again I'd update this to use that. Otherwise I'm making a fairly minimal first pass with this here.

TerryHowe replied:
Rather than make newReference public and being stuck with it, I was thinking in pkg/regsitry/plugin.go some function called GetPluginName(source string) (string, error) that does all this and uses newReference.

scottrigby replied:
WDYT? 4c83dd7

scottrigby committed:
New registry plugin func GetPluginName. Re-use regsitry.reference

Signed-off-by: Scott Rigby <scott@r6by.com>

TerryHowe reviewed (approved):
/lgtm

sabre1041 reviewed (dismissed):
Looks really good. Only one comment, but its not a blocking comment. Feel free to address your thoughts. Otherwise, LGTM

scottrigby commented on internal/plugin/installer/oci_installer.go:
```diff
@@ -0,0 +1,217 @@
+				return err
+			}
+
+			outFile, err := os.OpenFile(path, os.O_CREATE|os.O_RDWR, os.FileMode(header.Mode))
```
good catch. fixing

scottrigby committed:
Fix file handle management in tar extractors

Use defer outFile.Close() to avoid multiple close calls and ensure proper resource cleanup

Co-authored-by: Andrew Block <andy.block@gmail.com>
Signed-off-by: Scott Rigby <scott@r6by.com>

sabre1041 reviewed (approved):
LGTM
